metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/ConsensusService.scala

```diff
@@ -371,8 +422,15 @@ class ConsensusService[F[_]: Timer: Concurrent, N, A <: Agreement: Block](
 
       case CreateBlock(viewNumber, highQC) =>
         // Ask the application to create a block for us.
-        // TODO (PM-3109): Create block.
-        ???
+        appService.createBlock(highQC).flatMap {
+          case None =>
```

Comment:
In what situation is `None` an acceptable response?

Reply:
There can be a couple of reasons:
* A Prepare Q.C. means that a quorum of the federation has agreed that the BFT block pointed to by the High Q.C. is valid, confirmed by their Interpreters. But that doesn't mean that _every_ federation member had the data available to validate the block: maybe one of them lives on the "other side" of the PoW network and some transaction in the BFT block is about something they haven't received yet. If such a node becomes the next leader, they may not be able to produce a valid block. Say in the context of checkpointing, the leader is given a BFT ledger that has a last checkpoint they don't have yet, and now they have the option to nominate the next candidate. They may want to wait some time for this block to arrive, but if we time out, we can just move on to the next leader.
* It is highly unlikely, but, since this is happening in the background, the node may experience a round timeout, and jump ahead to catch up with others, clearning out earlier blocks, in which case I suppose this High Q.C. could be obsolete.
* It is possible that the `lastExecutedBlockHash` hasn't caught up with the High Q.C. yet. I'd expect that it should be possible to figure out what the state would be and produce a block, but maybe it isn't, maybe the leader just wants to pass. Although it should be possible to await the result inside the `ApplicationService` since it's the one that will be asked to do the execution as well.
* Perhaps the PoW Interpreter node crashes and doesn't respond.

Reply:
Right, so it isn't a perfectly normal situation, so perhaps it should be traced? Or would you prefer to do that in `ApplicationService` where more context would be available?

Reply:
I agree that some feedback would be good here, but you are right that the `ApplicationService` could tell much more about what the reason was. If that's acceptable, it would be more convenient for me, otherwise I'd have to change a few tracers in the robot example where a new type would be missing.

Reply:
👍 I added a comment to the revelant ticket